### PR TITLE
Make metric suffixes optional, expose unmapped events

### PR DESCRIPTION
**Pull request by discordianfish**
I change the README to show all examples without the auto suffixes but mention that, for now, they still get added. I did that because I think in one of the next releases we should change the defaults: Even though it might break things, I believe all our exporters should be serve as idiomatic example which here isn't the case.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -57,28 +57,33 @@ Metrics that don't match any mapping in the configuration file are translated
...
 
-    StatsD timer   -> Prometheus summary (suffix `_timer`)        <-- indicates timer quantiles
-                   -> Prometheus counter (suffix `_timer_total`)  <-- indicates total time spent
-                   -> Prometheus counter (suffix `_timer_count`)  <-- indicates total number of timer events
+    StatsD timer   -> Prometheus summary                    <-- indicates timer quantiles
+                   -> Prometheus counter (suffix `_total`)  <-- indicates total time spent
+                   -> Prometheus counter (suffix `_count`)  <-- indicates total number of timer events
 
-An example mapping configuration:
+If `-statsd.add-suffix` is set, the exporter appends the metric type (gauge,
```
Comment (juliusv) [inline README.md]: Be more explicit/precise about what they can expect: "(`_gauge`, `_counter`, `_timer`)".

File: README.md [hunk 2]
```diff
@@ -57,28 +57,33 @@ Metrics that don't match any mapping in the configuration file are translated
...
+counter, timer) to the resulting metrics. This is enabled by default for
+backward compatibility but discouraged to use. Instead better explicitly define
```
Comment (juliusv) [inline README.md]: "Instead, it is better to explicitly define..."

File: README.md [hunk 3]
```diff
@@ -87,13 +92,13 @@ This would transform these example StatsD metrics into Prometheus metrics as
 follows:
 
     test.dispatcher.FooProcessor.send.success (counter)
```
Comment (juliusv) [inline README.md]: I think you can get rid of the "(counter)" and "(gauge)" comments now that they don't actually have an effect in the example anymore.

File: main.go [hunk 4]
```diff
@@ -30,6 +30,7 @@ var (
 	statsdListenAddress = flag.String("statsd.listen-address", ":9125", "The UDP address on which to receive statsd metric lines.")
 	mappingConfig       = flag.String("statsd.mapping-config", "", "Metric mapping configuration file name.")
 	readBuffer          = flag.Int("statsd.read-buffer", 0, "Size (in bytes) of the operating system's transmit read buffer associated with the UDP connection. Please make sure the kernel parameters net.core.rmem_max is set to a value greater than the value specified.")
+	addSuffix           = flag.Bool("statsd.add-suffix", true, "Add the metric type (counter/gauge/timer) as suffix to the generated Prometheus metric (NOT recommended).")
```
Comment (juliusv) [inline main.go]: Maybe say "(NOT recommended, but set by default for backward compatibility)"?

File: main.go [hunk 5]
```diff
@@ -108,6 +109,9 @@ func watchConfig(fileName string, mapper *metricMapper) {
 func main() {
 	flag.Parse()
 
+	if *addSuffix {
+		log.Println("Warning: Using -statsd.add-suffix is discouraged. Better explicitly name metrics in config according to best pratices.")
```
Comment (juliusv) [inline main.go]: Instead of "Better..." I'd go with "We recommend explicitly naming metrics appropriately in the mapping configuration."

File: exporter.go [hunk 6]
```diff
@@ -189,6 +189,14 @@ func escapeMetricName(metricName string) string {
 	return metricName
 }
 
+func suffix(metricName, suffix string) string {
+	str := metricName
+	if *addSuffix {
```
Comment (juliusv) [inline exporter.go]: Let's not make this code base even uglier by directly reading flag values from here.

Make it an `Exporter` method instead and pass the desired value into `NewExporter`?

File: telemetry.go [hunk 7]
```diff
@@ -25,6 +25,10 @@ var (
 		},
 		[]string{"type"},
 	)
+	eventsMissCount = prometheus.NewCounter(prometheus.CounterOpts{
+		Name: "statsd_exporter_events_misses_total",
```
Comment (juliusv) [inline telemetry.go]: I'd go for `statsd_exporter_unmapped_events_total`, easier to understand what it's about.

Comment (juliusv) [inline telemetry.go]: (and then the var name should be renamed accordingly)

File: telemetry.go [hunk 8]
```diff
@@ -40,8 +44,8 @@ var (
 		[]string{"outcome"},
 	)
 	mappingsCount = prometheus.NewGauge(prometheus.GaugeOpts{
-		Name: "statsd_exporter_loaded_mappings_count",
-		Help: "The number of configured metric mappings.",
+		Name: "statsd_exporter_loaded_mappings",
```
Comment (juliusv) [inline telemetry.go]: While this is a better metric name and I support it, can we keep this breaking unrelated change out of this PR? Otherwise we will probably not catch it for the Changelog entry and all.

**Comment by juliusv** [comment]
:+1: otherwise

File: exporter.go [hunk 9]
```diff
@@ -209,13 +218,14 @@ func (b *Exporter) Listen(e <-chan Events) {
...
 			} else {
+				eventsUnmapped.Inc()
 				metricName = escapeMetricName(event.MetricName())
 			}
 
 			switch event.(type) {
 			case *CounterEvent:
 				counter := b.Counters.Get(
-					metricName+"_counter",
+					b.suffix(metricName, "counter"),
```
Comment (SuperQ) [inline exporter.go]: While we're fixing this, can we change this to `total` to match Prometheus metric name guidelines?

Comment (juliusv) [inline exporter.go]: @SuperQ The whole idea of introducing the `-statsd.add-suffix` flag (instead of just always omitting automatic suffixes) was to not have a breaking change. Changing the suffix to `_total` here would be breaking though.

**Comment by juliusv** [comment]
:+1:
